File: data_fetcher/csv_parser.py
import csv
import io

def parse_win_probability(raw_str):
    if not raw_str:
        return 0.5
    clean = str(raw_str).replace('%', '').strip()
    try:
        val = float(clean)
    except ValueError:
        return 0.5

    # Case 1: Decimal probability 0.0 - 1.0 (e.g. 0.37 or 0.626)
    if 0.0 < val < 1.0:
        return val

    # Case 2: Percentage string (e.g. 62.6% -> 0.626)
    if 1.0 <= val <= 99.9 and not str(raw_str).startswith('+') and not str(raw_str).startswith('-'):
        return val / 100.0

    # Case 3: American Odds (-150, +120, 100)
    if val < 0:
        prob = (-val) / (-val + 100.0)
    elif val > 0:
        prob = 100.0 / (val + 100.0)
    else:
        prob = 0.5

    return max(0.001, min(0.999, prob))

def parse_csv_content(csv_text):
    if not csv_text or not csv_text.strip():
        return []

    stream = io.StringIO(csv_text.strip())
    reader = csv.reader(stream)

    rows = []
    headers = []

    for i, row in enumerate(reader):
        if not row or not any(row):
            continue
        
        row_str = " ".join(row).upper()
        
        # Header detection
        if 'LEAGUE' in row_str or 'MARKET' in row_str or 'WIN %' in row_str or 'MODEL_PROB' in row_str:
            headers = [h.strip().upper() for h in row]
            continue

        if not headers:
            # 11-column default spec header if missing
            headers = ['LEAGUE', 'DATE', 'HOME', 'AWAY', 'DOUBLEHEADER', 'SECTION', 'MARKET', 'SELECTOR', 'POINT', 'SIDE', 'WIN %']

        get_val = lambda keyword: row[headers.index(keyword)].strip() if keyword in headers and headers.index(keyword) < len(row) else ''

        league = get_val('LEAGUE') or 'MLB'
        date = get_val('DATE') or '20250410'
        home = get_val('HOME')
        away = get_val('AWAY')
        doubleheader = get_val('DOUBLEHEADER') or '0'
        section = get_val('SECTION') or 'head_to_head'
        market = get_val('MARKET') or 'h2h'
        selector = get_val('SELECTOR')
        point = get_val('POINT')
        side = get_val('SIDE') or 'home'
        
        win_str = get_val('WIN %') or get_val('MODEL_PROB') or get_val('WIN') or get_val('PROB')
        model_prob = parse_win_probability(win_str)

        # Handle team_prop / player_prop selector fallback if HOME/AWAY are blank
        if section in ['team_prop', 'player_prop'] and selector and not home:
            home = selector

        # Un-slugify selector (e.g. joe-burrow -> Joe Burrow)
        clean_selector = selector.replace('-', ' ').title() if ('-' in selector and not selector.startswith('+') and not selector.startswith('-')) else selector

        # Un-slugify market (e.g. player-passing_yards-ou -> passing_yards)
        clean_market = market.lower().replace('-', '_').replace('player_', '').replace('_ou', '')

        rows.append({
            'id': f"row-{i}-{league}-{date}-{home}-{away}-{clean_market}-{side}-{point}",
            'league': league.upper(),
            'date': date,
            'home': home,
            'away': away,
            'doubleheader': doubleheader,
            'section': section.lower(),
            'market': clean_market,
            'raw_market': market,
            'selector': clean_selector,
            'raw_selector': selector,
            'point': point,
            'side': side.lower(),
            'win_str': win_str,
            'model_prob': model_prob,
        })

    return rows

File: src/services/csvParser.ts
import type { EightRainRow } from '../types';

export function parseWinProbability(winPctStr: string): { decimalProb: number; americanOdds: number } {
  if (!winPctStr) return { decimalProb: 0.5, americanOdds: -110 };
  
  const clean = winPctStr.trim().replace('%', '');
  const num = parseFloat(clean);
  
  if (isNaN(num)) return { decimalProb: 0.5, americanOdds: -110 };

  // Case 1: Decimal probability between 0 and 1 (e.g. 0.626)
  if (num > 0 && num < 1) {
    const dec = num;
    const american = decToAmericanOdds(dec);
    return { decimalProb: dec, americanOdds: american };
  }

  // Case 2: Percentage (e.g. 62.6 -> 0.626)
  if (num >= 1 && num <= 99.9 && !winPctStr.startsWith('+') && !winPctStr.startsWith('-') && !Number.isInteger(num)) {
    const dec = num / 100;
    const american = decToAmericanOdds(dec);
    return { decimalProb: dec, americanOdds: american };
  }

  // Case 3: American Odds (-150, +120)
  let decProb = 0.5;
  let american = num;
  if (num < 0) {
    decProb = (-num) / (-num + 100);
  } else if (num > 0) {
    decProb = 100 / (num + 100);
  }

  decProb = Math.max(0.001, Math.min(0.999, decProb));
  return { decimalProb: decProb, americanOdds: Math.round(american) };
}

export function decToAmericanOdds(p: number): number {
  if (p <= 0.001) return 10000;
  if (p >= 0.999) return -10000;
  if (p >= 0.5) {
    return Math.round(-100 * (p / (1 - p)));
  } else {
    return Math.round(100 * ((1 - p) / p));
  }
}

export function parse8rainCsv(csvText: string): EightRainRow[] {
  const lines = csvText.split(/\r?\n/).filter(line => line.trim().length > 0);
  if (lines.length === 0) return [];

  const rows: EightRainRow[] = [];

  // Find header row
  const headerIdx = lines.findIndex(l => {
    const u = l.toUpperCase();
    return u.includes('LEAGUE') && (u.includes('MARKET') || u.includes('HOME'));
  });

  if (headerIdx < 0) return [];

  const headers = parseCsvRow(lines[headerIdx]).map(h => h.toUpperCase().trim());

  // Find column mapping dynamically
  const colMap = {
    league: headers.findIndex(h => h.includes('LEAGUE')),
    date: headers.findIndex(h => h.includes('DATE')),
    home: headers.findIndex(h => h.includes('HOME')),
    away: headers.findIndex(h => h.includes('AWAY')),
    doubleheader: headers.findIndex(h => h.includes('DOUBLEHEADER')),
    section: headers.findIndex(h => h.includes('SECTION')),
    market: headers.findIndex(h => h.includes('MARKET')),
    selector: headers.findIndex(h => h.includes('SELECTOR')),
    point: headers.findIndex(h => h.includes('POINT')),
    side: headers.findIndex(h => h.includes('SIDE')),
    winPct: headers.findIndex(h => h.includes('MODEL_PROB') || h.includes('WIN') || h.includes('PROB')),
  };

  for (let i = headerIdx + 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line || line.startsWith('#') || line.startsWith('//')) continue;

    const cols = parseCsvRow(line);
    if (cols.length < 4) continue;

    const getCol = (idx: number) => (idx >= 0 && cols[idx] !== undefined) ? cols[idx].trim() : '';

    const league = getCol(colMap.league) || 'nfl';
    const date = getCol(colMap.date) || '20260927';
    const home = getCol(colMap.home);
    const away = getCol(colMap.away);
    const doubleheader = parseInt(getCol(colMap.doubleheader) || '0', 10) || 0;
    const section = getCol(colMap.section) || 'head_to_head';
    const market = getCol(colMap.market) || 'h2h';
    const selector = getCol(colMap.selector);
    const point = getCol(colMap.point);
    const side = getCol(colMap.side) || 'home';
    const winPctRaw = getCol(colMap.winPct);

    const { decimalProb, americanOdds } = parseWinProbability(winPctRaw);

    rows.push({
      id: `row-${i}-${league}-${home}-${away}-${market}-${side}`,
      league,
      date,
      home,
      away,
      doubleheader,
      section,
      market,
      selector,
      point,
      side,
      winPctRaw,
      winPctDecimal: decimalProb,
      americanOdds,
    });
  }

  return rows;
}

function parseCsvRow(row: string): string[] {
  const result: string[] = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < row.length; i++) {
    const char = row[i];
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      result.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }
  result.push(current.trim());
  return result;
}
